core/types/block.libevm.go

```diff
@@ -17,13 +17,74 @@
 package types
 
 import (
+	"fmt"
 	"io"
 
+	"github.com/ava-labs/libevm/libevm/pseudo"
 	"github.com/ava-labs/libevm/rlp"
 )
 
+// HeaderHooks are required for all types registered with [RegisterExtras] for
+// [Header] payloads.
+type HeaderHooks interface {
+	EncodeRLP(*Header, io.Writer) error
+	DecodeRLP(*Header, *rlp.Stream) error
+}
+
+var _ interface {
+	rlp.Encoder
+	rlp.Decoder
+} = (*Header)(nil)
+
+// EncodeRLP implements the [rlp.Encoder] interface.
 func (h *Header) EncodeRLP(w io.Writer) error {
+	if r := &registeredExtras; r.Registered() {
+		return r.Get().hooks.hooksFromHeader(h).EncodeRLP(h, w)
+	}
 	return h.encodeRLP(w)
 }
 
-var _ rlp.Encoder = (*Header)(nil)
+// decodeHeaderRLPDirectly bypasses the [Header.DecodeRLP] method to avoid
+// infinite recursion.
+func decodeHeaderRLPDirectly(h *Header, s *rlp.Stream) error {
+	type withoutMethods Header
+	return s.Decode((*withoutMethods)(h))
+}
+
+// DecodeRLP implements the [rlp.Decoder] interface.
+func (h *Header) DecodeRLP(s *rlp.Stream) error {
+	if r := &registeredExtras; r.Registered() {
+		return r.Get().hooks.hooksFromHeader(h).DecodeRLP(h, s)
+	}
+	return decodeHeaderRLPDirectly(h, s)
+}
+
+func (e ExtraPayloads[HPtr, SA]) hooksFromHeader(h *Header) HeaderHooks {
+	return e.Header.Get(h)
+}
+
+func (h *Header) extraPayload() *pseudo.Type {
+	r := &registeredExtras
+	if !r.Registered() {
+		// See params.ChainConfig.extraPayload() for panic rationale.
+		panic(fmt.Sprintf("%T.extraPayload() called before RegisterExtras()", r))
+	}
+	if h.extra == nil {
+		h.extra = r.Get().newHeader()
+	}
+	return h.extra
+}
+
+// NOOPHeaderHooks implements [HeaderHooks] such that they are equivalent to
+// no type having been registered.
+type NOOPHeaderHooks struct{}
```

Comment:
Do we need this type exported? Or is it only used in `types_test` packaged tests in this directory? If so would it make sense to have it defined in a `types_test` packaged **test** file in this directory to avoid polluting the core/types namespace?